common/src/main/java/org/apache/apex/common/util/JarHelper.java

```diff
@@ -70,41 +74,108 @@ public static String createJar(String prefix, File dir, boolean deleteOnExit) th
+  }
+
+  /**
+   * Returns a full path to the jar-file that contains the given class and all full paths to dependent jar-files
+   * that are defined in the property "apex-dependencies" of the manifest of the root jar-file.
+   * If the class is an independent file the method makes jar file from the folder that contains the class
+   * @param clazz Class
+   * @param makeJarFromFolder True if the method should make jar from folder that contains the independent class
+   * @param addJarDependencies True if the method should include dependent jar files
+   * @return List of names of the jar-files
+   */
+  public List<String> getJars(Class<?> clazz, boolean makeJarFromFolder, boolean addJarDependencies)
+  {
+    CodeSource codeSource = clazz.getProtectionDomain().getCodeSource();
+    if (codeSource == null) {
+      return null;
+    }
+
+    List<String> list = new LinkedList<>();
+    URL location  =  codeSource.getLocation();
+    String jarPath = sourceToJar.get(location);
+
+    if (jarPath == null) {
+      try {
+        URLConnection conn = clazz.getResource(clazz.getSimpleName() + ".class").openConnection();
+
+        if (conn instanceof JarURLConnection) {
+          jarPath = ((JarURLConnection)conn).getJarFileURL().getFile();
+          if (addJarDependencies) {
+            getDependentJarsFromManifest((JarURLConnection)conn, jarPath, list);
+            // add the location of the jar-file to cache only if the dependent jars were added to the jar list
+            sourceToJar.put(location, jarPath);
           }
+        } else if (makeJarFromFolder && ("file".equals(location.getProtocol()))) {
+          jarPath = createJar("apex-", new File(conn.getURL().getFile()).getParentFile(), false);
+          sourceToJar.put(location, jarPath);
         } else {
-          throw new AssertionError("Cannot resolve jar file for " + jarClass + ". URL " + location);
+          logger.warn("The class {} was loaded from incorrect url connection {}", clazz.getCanonicalName(), conn);
+          return list;
         }
-        sourceToJar.put(location, jar);
-        logger.debug("added sourceLocation {} as {}", location, jar);
+      } catch (IOException ex) {
+        String className = clazz.getCanonicalName();
+        logger.error("Cannot resolve jar file for the class {}", className, ex);
+        throw new RuntimeException("Cannot resolve jar file for the class " + className, ex);
       }
-      if (jar == null) {
-        throw new AssertionError("Cannot resolve jar file for " + jarClass);
+    }
+
+    list.add(jarPath);
+    logger.debug("added sourceLocation {} as {}", location, jarPath);
+
+    return list;
+  }
+
+  /**
+   * Gets dependent jar-files from manifest
+   * @param conn Jar connection
+   * @param jarPath path to jar file
+   * @param list List of target jar-files
+   * @throws IOException
+   */
+  private void getDependentJarsFromManifest(JarURLConnection conn, String jarPath, List<String> list) throws IOException
```

Comment:
Please consider using JarFile instead of JarURLConnection
I would expect the getDependentJarsFromManifest to return a list of the dependent jars, instead of accepting a list.